### PR TITLE
HTCONDOR-1325  Create (relative) directories in transfer_output_remaps

### PR DESCRIPTION
**Pull request by Todd-L-Miller**
If an output file is remapped to a relative path, and that relative path does not exist, create it, subject to LIMIT_DIRECTORY_ACCESS.  Because we implemented output and error log names using `transfer_output_remaps`, this change allows them to specify nonexistent relative paths as well, but this is a good thing, e.g., `output = $(CLUSTER)/output.log`.

This PR also removes two warnings from unrelated code.

# HTCondor Pull Request Checklist for internal reviewers

- [ ] Verify that (GitHub thinks) the merge is clean. If it isn't, and you're confident you can resolve the conflicts, do so. Otherwise, send it back to the original developer.
- [ ] Verify that the related Jira ticket exists and has a target version number and that it is correct.
- [ ] Verify that the Jira ticket is in review status and is assigned to the reviewer.
- [ ] Verify that the Jira ticket (HTCONDOR-xxx) is mentioned at the beginning of the title. Edit it, if not
- [ ] Verify that the branch destination of the PR matches the target version of the ticket
- [ ] Check for correctness of change
- [ ] Check for regression test(s) of new features and bugfixes (if the feature doesn't require root)
- [ ] Check for documentation, if needed
- [ ] Check for version history, if needed
- [ ] Check BaTLab dashboard for successful build (https://batlab.chtc.wisc.edu/results/workspace.php) and test for either the PR or a workspace build by the developer that has the Jira ticket as a comment.
- [ ] Check that each commit message references the Jira ticket (HTCONDOR-xxx)

## After the above
- Hit the merge button if the pull request is approved and it is not a security patch (security changes require 2 additional reviews)
- If the pull request is approved, take the ticket out of review state
- Assign JIRA Ticket back to the developer

### CODE REVIEW

**Review by JaimeFrey — CHANGES_REQUESTED** [review]
Why do we need to re-initialize allow_shadow_access()? The only reason I can see is if an entry in the allow list is a directory that we're only now creating and it's specified as a non-canonical path. I suggest we ignore that case (and say those directories must exist when the shadow is born).
Removing all of the directories in the relative path is not good. It can end up removing directories we didn't create, which may include ones outside of the allow_shadow_access() list. The other easy option is to leave all directories we created in place, which is also not great. To do this properly, we'd want a variant of realpath() that says "what would the canonical path be if we created any missing parent directories".

File: src/condor_utils/file_transfer.cpp [hunk 1]
```diff
@@ -2121,6 +2122,27 @@ FileTransfer::AddDownloadFilenameRemaps(char const *remaps) {
...
+		std::string path;
+		formatstr( path, "%s%c%s", iwd.c_str(), DIR_DELIM_CHAR, buffer.c_str() );
+		int rv = rmdir( path.c_str() );
+
+		if( rv == -1 ) {
+			dprintf( D_ALWAYS, "Failed to remove directory (%s) created outside of LIMIT_DIRECTORY_ACCESS (%d: %s), ignoring.\n", path.c_str(), errno, strerror(errno) );
+		}
+
+		idx = buffer.rfind( DIR_DELIM_CHAR );
+		buffer = buffer.substr(0, idx);
```
[the remark this answers is not in the document]
Comment (Todd-L-Miller) [reply src/condor_utils/file_transfer.cpp:2140]: The C++ spec defines `substr(0, idx)` as the whole string.  Is the thought that `erase()` is faster?

[the remark this answers is not in the document]
Comment (JaimeFrey) [reply src/condor_utils/file_transfer.cpp:2140]: Correct. erase() can edit the buffer in-place. substr() + assign probably involves copying the substring back and forth.

[the remark this answers is not in the document]
Comment (Todd-L-Miller) [reply src/condor_utils/file_transfer.cpp:2140]: The code looks a little clumsier to me as a result, but that's fine.

[the remark this answers is not in the document]
Comment (JaimeFrey) [reply src/condor_utils/file_transfer.cpp]: Agreed that it looks a little clumsier. I had hoped that erase(npos) would result in erasing nothing. But it apparently throws an exception.

**Comment by Todd-L-Miller** [comment]
> Why do we need to re-initialize allow_shadow_access()? The only reason I can see is if an entry in the allow list is a directory that we're only now creating and it's specified as a non-canonical path. I suggest we ignore that case (and say those directories must exist when the shadow is born).

Because it was necessary for a test I ran.  To be specific, the problem is not non-canonical paths, but paths that don't exist at shadow start-up but are named in the allow list (e.g., `/home/<username>/condor-output-directory`).

**Comment by Todd-L-Miller** [comment]
Jaime and I agreed in discussion that a better approach would be to extend/replace `mkdir_and_parents_if_needed()` with a function that was shadow-access aware and could therefore stop before creating an invalid directory.  If -- and only if -- it creates a directory would it need to update the shadow access cache.

File: src/condor_utils/file_transfer.cpp [hunk 2]
```diff
@@ -2150,28 +2151,125 @@ FileTransfer::AddDownloadFilenameRemaps(char const *remaps) {
...
+	if(! path.has_root_path()) {
+		dprintf( D_ALWAYS, "Internal logic error: shadow_safe_mkdir() called with relative path.  Refusing to make the directory.\n" );
+		return false;
+	}
+
+	// If the path already exists, we have nothing to do.
+	if( std::filesystem::exists( path ) ) { return true; }
+
+
+	if( priv != PRIV_UNKNOWN ) {
```
Comment (JaimeFrey) [inline src/condor_utils/file_transfer.cpp:2260]: The priv switch needs to happen before the exists() call.
TemporaryPrivSentry is how the cool kids ensure we restore the original priv state by the end of the function.

**Review by JaimeFrey — CHANGES_REQUESTED** [review]
Other than the comments I made on shadow_safe_mkdir(), this looks good.

File: src/condor_utils/file_transfer.cpp (hunk 2 above)
Comment (Todd-L-Miller) [reply src/condor_utils/file_transfer.cpp]: New commit makes these changes.

**Review by JaimeFrey — APPROVED** [review]
Looks good now.

**Pull request merged**